### PR TITLE
feat(medusa): Nested Categories Admin Update Endpoint

### PR DESCRIPTION
**Pull request by riqwan**
What:

Introduces an admin endpoint that allows a user to update a product category

Why:

This is part of a greater goal of allowing products to be added to multiple categories.

How:

- Creates a route on the admin scope to update category
- Creates a method in product category services to update a category

RESOLVES CORE-956

### CODE REVIEW

File: packages/medusa/src/services/product-category.ts [hunk 1]
```diff
@@ -116,6 +119,33 @@ class ProductCategoryService extends TransactionBaseService {
...
+
+      const productCategory = await this.retrieve(productCategoryId)
+
+      for (const key in productCategoryInput) {
+        if (isDefined(productCategoryInput[key])) {
+          productCategory[key] = productCategoryInput[key]
+        }
+      }
+
+      return productCategoryRepo.save(productCategory)
```
Comment (olivermrbl) [inline packages/medusa/src/services/product-category.ts:145]: todo: Add missing `await`
```suggestion
      return await productCategoryRepo.save(productCategory)
```

**Review by olivermrbl — CHANGES_REQUESTED** [review]
Strong work! Only a minor todo 💯 

Also, think @pKorsholm's suggestion to add events is great. For update and create :)

File: packages/medusa/src/api/routes/admin/product-categories/update-product-category.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
+import { IsOptional, IsString } from "class-validator"
+import { Request, Response } from "express"
+import { EntityManager } from "typeorm"
+
+import { ProductCategoryService } from "../../../../services"
+import { AdminProductCategoriesReqBase } from "../../../../types/product-category"
+import { FindParams } from "../../../../types/common"
+/**
+ * @oas [post] /product-categories/{id}
+ * operationId: "PostProductCategoriesCategory"
```
Comment (pKorsholm) [inline packages/medusa/src/api/routes/admin/product-categories/update-product-category.ts:10]: question: should this be `Category` or `ProductCategory`
```suggestion
 * operationId: "PostProductCategoriesProductCategory"
```

**Comment by olivermrbl** [comment]
Also, missing changeset :)

File: packages/medusa/src/api/routes/admin/product-categories/update-product-category.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+ *     type: boolean
+ *     description: A flag to make product category an internal category for admins
+ *   is_active:
+ *     type: boolean
+ *     description: A flag to make product category visible/hidden in the store front
+ *   parent_category_id:
+ *     type: string
+ *     description: The ID of the parent product category
+ */
+export class AdminPostProductCategoriesCategoryReq extends AdminProductCategoriesReqBase {
```
Comment (pKorsholm) [inline packages/medusa/src/api/routes/admin/product-categories/update-product-category.ts:116]:
```suggestion
// eslint-disable-next-line max-len
export class AdminPostProductCategoriesCategoryReq extends AdminProductCategoriesReqBase {
```

**Review by olivermrbl — APPROVED** [review]
LGTM!

File: packages/medusa/src/services/product-category.ts [hunk 4]
```diff
@@ -116,6 +119,33 @@ class ProductCategoryService extends TransactionBaseService {
...
+        this.productCategoryRepo_
+      )
+
+      const productCategory = await this.retrieve(productCategoryId)
+
+      for (const key in productCategoryInput) {
+        if (isDefined(productCategoryInput[key])) {
+          productCategory[key] = productCategoryInput[key]
+        }
+      }
```
Comment (pKorsholm) [inline packages/medusa/src/services/product-category.ts:143]: suggestion(non-blocking): The repositories can merge entities as so: 
```suggestion
      productCategoryRepo.merge(productCategory, productCategoryInput)
```

**Review by pKorsholm — APPROVED** [review]
LGTM!

Only have a couple of non-blocking nit-picks 😄

File: packages/medusa/src/api/routes/admin/product-categories/update-product-category.ts (hunk 2 above)
Comment (riqwan) [reply packages/medusa/src/api/routes/admin/product-categories/update-product-category.ts:10]: Based on the code conventions doc, it says this -

```
If a resource consists of multiple words you don’t have to repeat both in sub-segments. E.g.:

`GET /admin/sales-channels/:id` → `GetSalesChannelsChannel`
```

File: integration-tests/api/__tests__/admin/product-category.ts [hunk 5]
```diff
@@ -27,6 +28,7 @@ describe("/admin/product-categories", () => {
     const [process, connection] = await startServerWithEnvironment({
       cwd,
       env: { MEDUSA_FF_PRODUCT_CATEGORIES: true },
+      verbose: true
```
[the remark this answers is not in the document]
Comment (riqwan) [reply integration-tests/api/__tests__/admin/product-category.ts]: this was removed a while ago, how did you still have this in your view :D

File: packages/medusa/src/services/product-category.ts (hunk 4 above)
Comment (riqwan) [reply packages/medusa/src/services/product-category.ts:143]: nice, didn't know this!

**Review by adrien2p — APPROVED** [review]
LGTM

File: packages/medusa/src/services/product-category.ts [hunk 6]
```diff
@@ -116,6 +119,29 @@ class ProductCategoryService extends TransactionBaseService {
...
+    productCategoryInput: UpdateProductCategoryInput
+  ): Promise<ProductCategory> {
+    return await this.atomicPhase_(async (manager) => {
+      const productCategoryRepo = manager.getCustomRepository(
+        this.productCategoryRepo_
+      )
+
+      const productCategory = await this.retrieve(productCategoryId)
+
+      productCategoryRepo.merge(productCategory, productCategoryInput)
```
Comment (olivermrbl) [inline packages/medusa/src/services/product-category.ts:139]: comment: Haven't seen `merge` before, but I kinda like it. That said, should we stick with the `for each` for consistency?

Comment (olivermrbl) [reply packages/medusa/src/services/product-category.ts:139]: If we go with merge, I believe we have to correct the unit tests :)

Comment (adrien2p) [reply packages/medusa/src/services/product-category.ts:139]: Are you sure? merge will add both props to the new output entity, so only the props from the inputs will get updated

Comment (riqwan) [reply packages/medusa/src/services/product-category.ts]: reverted back to the for loop. I feel weird re-writing merge logic inside the repository mocks 😄

**Pull request merged**